server/src/main/java/org/opensearch/search/backpressure/SearchBackpressureService.java

```diff
@@ -130,18 +142,29 @@ void doRun() {
             return;
         }
 
-        // We are only targeting in-flight cancellation of SearchShardTask for now.
-        List<SearchShardTask> searchShardTasks = getSearchShardTasks();
+        List<CancellableTask> searchTasks = getSearchTasks();
+        List<CancellableTask> searchShardTasks = getSearchShardTasks();
+        List<CancellableTask> cancellableTasks = new ArrayList<>();
 
         // Force-refresh usage stats of these tasks before making a cancellation decision.
+        taskResourceTrackingService.refreshResourceStats(searchTasks.toArray(new Task[0]));
         taskResourceTrackingService.refreshResourceStats(searchShardTasks.toArray(new Task[0]));
 
-        // Skip cancellation if the increase in heap usage is not due to search requests.
-        if (isHeapUsageDominatedBySearch(searchShardTasks) == false) {
+        // Check if increase in heap usage is due to SearchTasks
+        if (isHeapUsageDominatedBySearch(searchTasks, getSettings().getSearchTaskSettings().getTotalHeapBytesThreshold())) {
+            cancellableTasks.addAll(searchTasks);
+        }
+
+        // Check if increase in heap usage is due to SearchShardTasks
+        if (isHeapUsageDominatedBySearch(searchShardTasks, getSettings().getSearchShardTaskSettings().getTotalHeapBytesThreshold())) {
+            cancellableTasks.addAll(searchShardTasks);
+        }
+
+        if (cancellableTasks.isEmpty()) {
             return;
         }
```

Comment:
The intent here was to validate if node duress is due to search requests in general (including both search tasks and shard tasks). Do we see additional value in splitting these thresholds?

Is it possible that sum of search task and shard task usage is high, but individually they don't exceed their respective thresholds. Should this be something to handle?

Reply:
Well, I want to make sure that SearchTasks do not get cancelled when node is in duress due to high resource usage by the SearchShardTasks, or vice versa. We also have 2 different dynamic settings called `search_backpressure.search_task.total_heap_percent_threshold` and `search_backpressure.search_shard_task.total_heap_percent_threshold` to define the heap percentage thresholds for both the task types, so these will decide whether the heap usage is dominated by any of the task type.

> Is it possible that sum of search task and shard task usage is high, but individually they don't exceed their respective thresholds. Should this be something to handle?

In this case(only when `cancellableTasks` is empty after both the checks), should we add all the SearchTasks and SearchShardTasks in the `cancellableTasks` list and let the cancellation score take care of the cancellation priority?
Also, I would like to keep the `isHeapUsageDominatedBySearch()` check for both the task types as it is, and only when `cancellableTasks` is empty after the checks, we will add all the tasks since node is in duress.

Reply:
@nssuresh2007 , having combined thresholds makes things less complex and implementation will be easy. However, having separate thresholds for each task type gives better control to customers. By having separate thresholds, we can also make sure that SearchTasks do not get cancelled when node is in duress due to high resource consumption by the SearchShardTasks, or vice versa. 
Consider a scenario, in a single node cluster, a complex query comes in where SearchShardTask has to do a lot of processing and SearchTask is just waiting for the SearchShardTask to return the result. If we have combined thresholds, SearchTask will also be marked for cancellation, even if it is not consuming so many resources. This kind of scenarios will be handled when we have separate thresholds.
One benefit on having the combined threshold is that the scenario mentioned by @ketanv3 in previous comment will be handled. However, I have suggested a solution for that as well in the above comment.